stitches/dev_matchup.py

```diff
@@ -8,155 +8,283 @@
-    :param size:                    An integer value for the size of the window to use when calculating the rolling mean
-    :type size:                     int
+  if not (req_exps.issubset(exp_set)):
+    raise TypeError(f'{input_data} is missing data for the historical experiment.')
 
-    :return:                   A pandas data frame of the smoothed time series (rolling mean applied)
-    """
-    # TODO add some code that checks the inputs of the input_data
-    # inputs
-    req_cols = {'model', 'experiment', 'ensemble', 'year', 'value', 'variable'}
-    col_names = set(input_data.columns)
-    if not (req_cols.issubset(col_names)):
-        raise TypeError(f'Missing columns from "{input_data}".')
-
-    req_exps = {"historical"}
-    exp_set = set(input_data['experiment'].unique())
-
-    if not (req_exps.issubset(exp_set)):
-        raise TypeError(f'{input_data} is missing data for the historical experiment.')
-
-    # Relabel the historical values so that there is a continuous rolling mean between the
-    # historical and future values.
-    # #######################################################
-    # Subset the historical data
-    historical_data = input_data[input_data["experiment"] == "historical"].copy()
-
-    # Create a subset of the future data
-    future_data = input_data[input_data["experiment"] != "historical"].copy()
-    future_scns = set(future_data["experiment"].unique())
-
-    frames = []
-    for scn in future_scns:
-        d = historical_data.copy()
-        d["experiment"] = scn
-        frames.append(d)
-
-    frames.append(future_data)
-    data = pd.concat(frames)
-
-    # Index the data frame by the year, so that the rolling mean respects the years
-    data.index = data['year']
-
-    # Now calculate the rolling mean that uses a centered window.
-    group_by = ['model', 'experiment', 'ensemble', 'variable']
-    rslt = data.groupby(group_by)['value'].rolling(size, center=True).mean().reset_index()
-
-    return rslt
+  # Relabel the historical values so that there is a continuous rolling mean between the
+  # historical and future values.
+  # #######################################################
+  # Subset the historical data
+  historical_data = input_data[input_data["experiment"] == "historical"].copy()
+
+  # Create a subset of the future data
+  future_data = input_data[input_data["experiment"] != "historical"].copy()
+  future_scns = set(future_data["experiment"].unique())
+
+  frames = []
+  for scn in future_scns:
+    d = historical_data.copy()
+    d["experiment"] = scn
+    frames.append(d)
+
+  frames.append(future_data)
+  data = pd.concat(frames)
+
+  # Index the data frame by the year, so that the rolling mean respects the years
+  data.index = data['year']
+
+  # Now calculate the rolling mean that uses a centered window.
+  group_by = ['model', 'experiment', 'ensemble', 'variable']
+  rslt = data.groupby(group_by)['value'].rolling(size, center=True).mean().reset_index()
```

Comment:
I wonder if we want/need to handle the starting and ending years so that they don't just get replaced with NaN. If we end up with a 20-ish year smoothing window, that takes out the first and last 10 years of the data set. Losing 1850-1860 is probably fine for our applications but 2090-2100 is tougher to have no data for. In the past, I've written custom rolling average functions in R so that when you have those end years with not enough data, you just average what you do have. So for a centered average of 2095 with a 20 year window, you'd take the average from 2086-2100. It's imperfect but you keep all the years at least. @kdorheim  do you think this is something we want? if so, I can try to translate my R function into python here? Just something I thought of as I've been digging more into the smoothing investigation. I realized with the plots I've made where I have x=window size and y = error, the window=5 years is comparing 80 years of ESM data to 80 years of reference, but the window=50 years is only comparing 35 years of ESM data to 35 years of reference.

Reply:
For the record, @abigailsnyder has implemented this in the branch `acs-smoothing-investigation` which will eventually be merged in here